Method annotated @WithoutTenant throws "No Session found" error

1. Define a GORM Data Service in a Multi-Tenant Application
2. Annotate a method in the service with `@WithoutTenant`
3. Attempt to call annotated method

### Expected Behaviour

Method should return the expected result

### Actual Behaviour

```
Caused by: org.springframework.dao.DataAccessResourceFailureException: Could not obtain current Hibernate Session; nested exception is org.hibernate.HibernateException: No Session found for current thread
	at org.grails.orm.hibernate.GrailsHibernateTemplate.getSession(GrailsHibernateTemplate.java:337)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.doExecute(GrailsHibernateTemplate.java:286)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:243)
	at org.grails.orm.hibernate.GrailsHibernateTemplate.execute(GrailsHibernateTemplate.java:117)
	at org.grails.orm.hibernate.AbstractHibernateDatastore.withSession(AbstractHibernateDatastore.java:355)
	at grails.gorm.multitenancy.Tenants.withoutId_closure1(Tenants.groovy:204)
	at grails.gorm.multitenancy.Tenants$CurrentTenant.withoutTenant(Tenants.groovy:380)
	at grails.gorm.multitenancy.Tenants.withoutId(Tenants.groovy:198)
	at org.grails.datastore.gorm.services.DefaultTenantService.withoutId(DefaultTenantService.groovy:59)
	... 1 more
```

### Environment Information

- **Operating System**: macOS
- **GORM Version:** 6.1.9 (6.1.10 and 6.1.11 are also affected)
- **Grails Version (if using Grails):** N/A
- **JDK Version:** 1.8

### Example Application

- https://github.com/ZacharyKlein/without-tenant-issue

--- a/grails-datastore-gorm/src/main/groovy/grails/gorm/multitenancy/Tenants.groovy
+++ b/grails-datastore-gorm/src/main/groovy/grails/gorm/multitenancy/Tenants.groovy
@@ -199,7 +199,7 @@ class Tenants {
             if (multiTenantCapableDatastore.getMultiTenancyMode().isSharedConnection()) {
                 def i = callable.parameterTypes.length
                 if(i == 0 ) {
-                    throw new IllegalArgumentException("Provided closure accepts too many arguments")
+                    return callable.call()
                 } else {
                     return multiTenantCapableDatastore.withSession { session ->
                         return callable.call(session)
@@ -272,7 +272,7 @@ class Tenants {
 
                 }
             }
-        }
+        } as T
     }
 
     /**

--- a/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/multitenancy/transform/TenantTransform.groovy
+++ b/grails-datastore-gorm/src/main/groovy/org/grails/datastore/gorm/multitenancy/transform/TenantTransform.groovy
@@ -41,7 +41,6 @@ import static org.codehaus.groovy.ast.ClassHelper.CLOSURE_TYPE
 import static org.codehaus.groovy.ast.ClassHelper.make
 import static org.codehaus.groovy.ast.tools.GeneralUtils.*
 import static org.grails.datastore.gorm.transform.AstMethodDispatchUtils.callD
-import static org.grails.datastore.mapping.reflect.AstUtils.ZERO_PARAMETERS
 import static org.grails.datastore.mapping.reflect.AstUtils.copyParameters
 import static org.grails.datastore.mapping.reflect.AstUtils.varThis
 
@@ -70,6 +69,8 @@ class TenantTransform extends AbstractDatastoreMethodDecoratingTransformation im
      */
     public static final int POSITION = TransactionalTransform.POSITION - 100
 
+    private static final Parameter[] N0_PARAMETER = null
+
     @Override
     protected String getRenamedMethodPrefix() {
         return RENAMED_METHOD_PREFIX
@@ -87,13 +88,11 @@ class TenantTransform extends AbstractDatastoreMethodDecoratingTransformation im
 
         ClassNode serializableClassNode = make(Serializable)
         ClassNode annotationClassNode = annotationNode.classNode
-        if(CURRENT_TENANT_ANNOTATION_TYPE.equals(annotationClassNode)) {
-            return makeDelegatingClosureCall( tenantServiceVar, "withCurrent", params( param(serializableClassNode, VAR_TENANT_ID)), originalMethodCallExpr, variableScope)
-        }
-        else if(WITHOUT_TENANT_ANNOTATION_TYPE.equals(annotationClassNode)) {
-            return makeDelegatingClosureCall( tenantServiceVar, "withoutId", ZERO_PARAMETERS, originalMethodCallExpr, variableScope)
-        }
-        else {
+        if (CURRENT_TENANT_ANNOTATION_TYPE.equals(annotationClassNode)) {
+            return makeDelegatingClosureCall(tenantServiceVar, "withCurrent", params(param(serializableClassNode, VAR_TENANT_ID)), originalMethodCallExpr, variableScope)
+        } else if (WITHOUT_TENANT_ANNOTATION_TYPE.equals(annotationClassNode)) {
+            return makeDelegatingClosureCall(tenantServiceVar, "withoutId", N0_PARAMETER, originalMethodCallExpr, variableScope)
+        } else {
             // must be @Tenant
             Expression annValue = annotationNode.getMember("value")
             if(annValue instanceof ClosureExpression) {

--- a/grails-datastore-gorm/src/test/groovy/grails/gorm/annotation/multitenancy/CurrentTenantTransformSpec.groovy
+++ b/grails-datastore-gorm/src/test/groovy/grails/gorm/annotation/multitenancy/CurrentTenantTransformSpec.groovy
@@ -1,7 +1,5 @@
 package grails.gorm.annotation.multitenancy
 
-import grails.gorm.multitenancy.TenantService
-import org.grails.datastore.mapping.core.Datastore
 import spock.lang.Specification
 
 /**
@@ -14,6 +12,103 @@ class CurrentTenantTransformSpec extends Specification {
         def bookService = new GroovyShell().evaluate('''
 import grails.gorm.multitenancy.CurrentTenant
 class BookService {
+    @CurrentTenant
+    List listBooks() {
+        return ["The Stand"]
+    }
+}
+new BookService()
+
+''')
+        when:"the list books method is invoked"
+        def result = bookService.listBooks()
+
+        then:"An exception was thrown because GORM is not setup"
+        thrown(IllegalStateException)
+
+    }
+
+    void "test @CurrentTenant transforms a service class and makes a method in current tenant handling"() {
+        given:"A service with @CurrentTenant applied as the class level"
+        def bookService = new GroovyShell().evaluate('''
+import grails.gorm.multitenancy.CurrentTenant
+
+@CurrentTenant
+class BookService {
+   
+    List listBooks() {
+        return ["The Stand"]
+    }
+}
+new BookService()
+
+''')
+        when:"the list books method is invoked"
+        def result = bookService.listBooks()
+
+        then:"An exception was thrown because GORM is not setup"
+        thrown(IllegalStateException)
+
+    }
+
+    void "test @CurrentTenant transforms a service class and a method marked with @WithoutTenant in no tenant handling"() {
+        given:"A service with @CurrentTenant applied as the class level"
+        def bookService = new GroovyShell().evaluate('''
+import grails.gorm.multitenancy.CurrentTenant
+import grails.gorm.multitenancy.WithoutTenant
+
+@CurrentTenant
+class BookService {
+   
+   @WithoutTenant
+    List listBooks() {
+        return ["The Stand"]
+    }
+}
+new BookService()
+
+''')
+        when:"the list books method is invoked"
+        def result = bookService.listBooks()
+
+        then:"An exception was thrown because GORM is not setup"
+        thrown(IllegalStateException)
+
+    }
+
+    void "test @WithoutTenant transforms a service class and makes a method that is wrapped in without tenant handling"() {
+        given:"A service with @CurrentTenant applied as the class level"
+        def bookService = new GroovyShell().evaluate('''
+import grails.gorm.multitenancy.CurrentTenant
+import grails.gorm.multitenancy.WithoutTenant
+
+@WithoutTenant
+class BookService {
+   
+    List listBooks() {
+        return ["The Stand"]
+    }
+}
+new BookService()
+
+''')
+        when:"the list books method is invoked"
+        def result = bookService.listBooks()
+
+        then:"An exception was thrown because GORM is not setup"
+        thrown(IllegalStateException)
+
+    }
+
+    void "test @WithoutTenant transforms a service class and a method marked with @CurrentTenant in current tenant handling"() {
+        given:"A service with @CurrentTenant applied as the class level"
+        def bookService = new GroovyShell().evaluate('''
+import grails.gorm.multitenancy.CurrentTenant
+import grails.gorm.multitenancy.WithoutTenant
+
+@WithoutTenant
+class BookService {
+   
     @CurrentTenant
     List listBooks() {
         return ["The Stand"]

